### PR TITLE
L2-596: Link existing canister

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

User can link an existing canister to the nns dapp user.

# Changes

* New Modal CreateOrLinkCanisterModal with two screens: SelectNewCanisterType and AttachCanister.
* New canister api: "attachCanister".
* New cansiter service: "attachCanister".
* Button in "Canisters" route opens CreateOrLinkCanisterModal.
* New UI Component: CardItem, created from cards in AddAccountType and reused there.
* New UI Component: PrincipalInput.

# Tests

* Test the Link Canister flow.
* Test the new canister api "attachCanister".
* Test the new canister service "attachCanister"
* Test that Canisters route opens CreateOrLinkCanisterModal.
* Test new UI Components.

### CODE REVIEW

**Comment by lmuntaner** [comment]
UX Video:

https://user-images.githubusercontent.com/4550653/170430964-c974773e-2805-47e2-9016-f56aae106f2f.mov

**Comment by lmuntaner** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/svelte/src/lib/components/ui/CardItem.svelte [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<script lang="ts">
+  export let title: string;
+  export let subtitle: string;
+  export let testId: string | undefined = undefined;
+</script>
+
+<div class="card-item" role="button" on:click data-tid={testId}>
+  <h4>{title}</h4>
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/CardItem.svelte:8]: use slots for title and subtitle

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  import { attachCanister } from "../../services/canisters.services";
+  import { createEventDispatcher } from "svelte";
+  import { toastsStore } from "../../stores/toasts.store";
+
+  let address: string = "";
+  let validPrincipal: Principal | undefined;
+  $: validPrincipal = getPrincipalFromString(address);
+  let showError: boolean = false;
+
+  const showErrorIfAny = () => {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:16]: we don't have this validator already somewhere else in the code?

if yes, can we avoid duplicate code and create a validator function?

File: frontend/svelte/src/lib/components/ui/CardItem.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/CardItem.svelte:8]: not really, I am still not used to them. I will do that.

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    showError = address.length > 0 && validPrincipal === undefined;
+  };
+  // Hide error on change
+  $: address, (showError = false);
+
+  const dispatcher = createEventDispatcher();
+  const attach = async () => {
+    // Button is only enabled when validPrincipal is defined
+    if (validPrincipal !== undefined) {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:25]: in accounts component I display an error even if I know there are few or no chances to reach that point if you want to follow the same pattern

```
if (icp === undefined) {
      toastsStore.error({
        labelKey: "error.amount_not_valid",
      });
      return;
    }
```

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 2 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:16]: yes, but it was simpler like this, I don't mind duplicating code. But I will refactor

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 3 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:25]: ok

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      startBusy({ initiator: "attach-canister" });
+      await attachCanister(validPrincipal);
+      toastsStore.success({
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:28]: the success message is displayed even if there is an error?

File: frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<script>
+  import { createEventDispatcher } from "svelte";
+  import { i18n } from "../../stores/i18n";
+  import CardItem from "../ui/CardItem.svelte";
+
+  const dispatcher = createEventDispatcher();
+  const selectCreate = () => {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte:7]: nth: refactor `selectCreate` and `selectLink` in a single function since you use the same dispatcher type in both

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 4 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:28]: oh, no, my bad. Initially I wanted to throw, but then I remebered we deal with this in the services.

File: frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte (hunk 5 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte:7]: ok, but they will be different afterwards, they will dispatch the same event with different data.

**Comment by peterpeterparker** [comment]
For UI consistency reason, can you center the label and make it look like a cta? something like

Capture d’écran 2022-05-26 à 08 46 08

File: frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte (hunk 5 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte:10]: done

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 4 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte:30]: done

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 3 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte]: done

File: frontend/svelte/src/lib/components/canisters/AttachCanister.svelte (hunk 2 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/AttachCanister.svelte]: done in the PrincipalInput

File: frontend/svelte/src/lib/components/ui/CardItem.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/CardItem.svelte]: done

**Comment by lmuntaner** [comment]
New screenshots:

Screenshot 2022-05-26 at 12 03 55
Screenshot 2022-05-26 at 12 04 04

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

File: frontend/svelte/src/lib/api/canisters.api.ts [hunk 6]
```diff
@@ -158,5 +179,7 @@ const canisters = async (
     agent,
   });
 
-  return { cmc, icMgt };
+  const { canister: nnsDapp } = await nnsDappCanister({ identity });
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/api/canisters.api.ts:182]: thinking about it now, can be solved in another PR, but instead of three `await` one after the other it would be more performant to `Promise.all`

```
const [cmc, icMgmt, nnsDapp] = await Promise.all([...]);
return {cmc, icMgmt, nnsDapp};
```

File: frontend/svelte/src/lib/components/accounts/AddAccountType.svelte [hunk 7]
```diff
@@ -17,55 +18,33 @@
...
-    h4 {
-      line-height: 1;
-      margin-bottom: var(--padding);
-    }
-
-    span {
-      color: var(--gray-200);
-    }
+    // Need to overwrite the default of wizrd-wrapper
+    justify-content: center !important;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/accounts/AddAccountType.svelte:48]: please do not use `!important` these can break a all layout

either a stronger selector or if not possible, pass a css variable

File: frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    <svelte:fragment slot="subtitle"
+      >{$i18n.canisters.link_canister_subtitle}</svelte:fragment
+    >
+  </CardItem>
+</div>
+
+<style lang="scss">
+  .wizard-wrapper {
+    // Need to overwrite the default of wizrd-wrapper
+    justify-content: center !important;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte:40]: likewise

File: frontend/svelte/src/lib/components/ui/CardItem.svelte [hunk 9]
```diff
@@ -0,0 +1,37 @@
+<script lang="ts">
+  export let testId: string | undefined = undefined;
+</script>
+
+<div class="card-item" role="button" on:click data-tid={testId}>
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/CardItem.svelte:5]: is it a `div` or should it be an `article` if we compare with our other cards?

File: frontend/svelte/src/lib/components/ui/PrincipalInput.svelte [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  import { i18n } from "../../stores/i18n";
+  import { getPrincipalFromString } from "../../utils/accounts.utils";
+  import InputWithError from "./InputWithError.svelte";
+
+  export let placeholderLabelKey: string;
+  export let name: string;
+  export let validPrincipal: Principal | undefined = undefined;
+
+  let address: string = "";
+  $: validPrincipal = getPrincipalFromString(address);
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:12]: shouldn't be:

```
$: validPrincipal = getPrincipalFromString(address) !== undefined
```

?

File: frontend/svelte/src/lib/components/accounts/AddAccountType.svelte (hunk 7 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/accounts/AddAccountType.svelte:48]: ok

File: frontend/svelte/src/lib/components/ui/CardItem.svelte (hunk 9 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/CardItem.svelte:5]: it was a `div` before, I didn't want to change it, maybe there was a reason.

But I can make it an article.

File: frontend/svelte/src/lib/components/ui/PrincipalInput.svelte (hunk 10 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:12]: I will rename `validPrincipal` to just `principal`, I was actually about to do it because I found it confusing as well.

File: frontend/svelte/src/lib/components/accounts/AddAccountType.svelte (hunk 7 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/accounts/AddAccountType.svelte]: done

File: frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte (hunk 8 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/canisters/SelectNewCanisterType.svelte]: done

File: frontend/svelte/src/lib/components/ui/PrincipalInput.svelte (hunk 10 above)
Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte]: done

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

File: frontend/svelte/src/lib/api/canisters.api.ts (hunk 6 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/api/canisters.api.ts:182]: "als agree", ja 👍😉

File: frontend/svelte/src/lib/components/ui/PrincipalInput.svelte [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+  export let placeholderLabelKey: string;
+  export let name: string;
+  export let principal: Principal | undefined = undefined;
+
+  let address: string = "";
+  $: principal = getPrincipalFromString(address);
+  let showError: boolean = false;
+
+  const showErrorIfAny = () => {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: any particular reason why you are creating a `principal` variable that has no particular other purpose that being use in the validation?

as currently coded `showErrorIfAny` is probably evaluated two times on each changes of the address where it can be evaluated just once

```
const showErrorIfAny = () => address.length > 0 && getPrincipalFromString(address) === undefined
```

but maybe you will need the variable later on anyway?

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks!

Please have a look to my last comment before merge.

File: frontend/svelte/src/lib/components/ui/PrincipalInput.svelte (hunk 11 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: also is `getPrincipalFromString` actually heavy on computation - i.e. should the error check be debounced? if so, for sure it can be done in another PR, just wanted to ask

Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: Now it could be easily done separately because it's on a single component.

I don't know a better way than trying to convert it to Principal to check whether a string is a valid principal or not.

An improvement could be to check only on blur for example. Something we can definitly improve in another PR.

Comment (peterpeterparker) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: I think you did not get my first comment but it's alright, thanks for the feedback.

Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: Mhm, well the `principal` variable is bound from the parent.

```
<PrincipalInput
      bind:principal
      placeholderLabelKey="canisters.canister_id"
      name="canister-principal"
    />
```

Comment (lmuntaner) [reply frontend/svelte/src/lib/components/ui/PrincipalInput.svelte:15]: this way it can easily get used there.
